crates/partition-store/src/snapshots/repository.rs

```diff
@@ -389,22 +482,302 @@ impl SnapshotRepository {
             })
             .unwrap_or_else(|| PutOptions::from(PutMode::Create));
 
-        // Note: this call may return an error on concurrent modification. Since we don't expect any
-        // contention (here), and this doesn't cause any correctness issues, we don't bother with
-        // retrying the entire put_snapshot attempt on object_store::Error::Precondition.
-        let put_result = self
-            .object_store
-            .put_opts(&latest_path, latest, conditions)
+        self.object_store
+            .put_opts(&latest_path, latest_payload, conditions)
             .await
             .map_err(|e| PutSnapshotError::from(e, progress.clone()))?;
 
         debug!(
-            etag = put_result.e_tag.unwrap_or_default(),
+            etag = "updated",
             key = ?latest_path,
             "Successfully updated latest snapshot pointer",
         );
 
-        Ok(())
+        if !new_latest.pending_deletions.is_empty() {
+            self.spawn_cleanup_task(snapshot.partition_id, new_latest.pending_deletions.clone());
+        }
+
+        Ok(ArchivedLsn::from(&new_latest))
+    }
+
+    fn determine_format_version(&self, current: Option<&LatestSnapshot>) -> LatestSnapshotVersion {
+        if self.retain_snapshots.is_some() {
+            if let Some(latest) = current
+                && latest.version == LatestSnapshotVersion::V1
+            {
+                debug!("Upgrading latest snapshot format from V1 to V2");
+            }
+            LatestSnapshotVersion::V2
+        } else {
+            current
+                .map(|l| l.version)
+                .unwrap_or(LatestSnapshotVersion::V1)
+        }
+    }
+
+    fn build_latest_v1(&self, snapshot: &PartitionSnapshotMetadata) -> LatestSnapshot {
+        LatestSnapshot::from_snapshot(snapshot)
+    }
+
+    fn build_latest_v2(
+        &self,
+        snapshot: &PartitionSnapshotMetadata,
+        current: Option<&LatestSnapshot>,
+    ) -> anyhow::Result<LatestSnapshot> {
+        let original_pending_deletions = current
+            .map(|l| l.pending_deletions.clone())
+            .unwrap_or_default();
+
+        if original_pending_deletions.len() >= MAX_PENDING_DELETIONS {
+            anyhow::bail!(
+                "Snapshot creation suspended: {} pending deletions exceed threshold {}. \
+                Check object store connectivity and permissions.",
+                original_pending_deletions.len(),
+                MAX_PENDING_DELETIONS
+            );
+        }
```

Comment:
Do we really want to stop creating snapshots if we go beyond the max pending deletions? This will stop trimming the log which increases the recovery time and disk usage. If we kept creating snapshots, the bad thing that can happen is S3 storage growth, which is not ideal but might be less severe than running out of disk on one of our log servers.